### PR TITLE
added a scope-check command

### PR DESCRIPTION
**Pull request by nanjekyejoannah**
# I have added a scope-check command.

### CODE REVIEW

File: scopecheck/test_scopecheck.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package scopecheck
```
Comment (jonasfj) [inline scopecheck/test_scopecheck.go]: test cases are awesome, let's name it `scopecheck_test.go` then it only gets compiled when running `go test ./...`

File: subtree_imports.go [hunk 2]
```diff
@@ -3,3 +3,5 @@ package main
 
 import _ "github.com/taskcluster/taskcluster-cli/apis"
 import _ "github.com/taskcluster/taskcluster-cli/extpoints"
+import _ "github.com/taskcluster/taskcluster-cli/scopecheck"
+import _ "github.com/taskcluster/taskcluster-cli/scope"
```
Comment (jonasfj) [inline subtree_imports.go]: probably a line too much here...

File: scopecheck/scopecheck.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+func usage() string {
+	return `Usage:
+  taskcluster scope-check <scope1> satifies <scope2>
+`
+}
+
+func (scopecheck) Usage() string {
+	return usage()
```
Comment (jonasfj) [inline scopecheck/scopecheck.go]: Why the helper method?

File: scopecheck/scopecheck.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+type Auth tcclient.ConnectionData
+
+func New(credentials *tcclient.Credentials) *Auth {
```
Comment (jonasfj) [inline scopecheck/scopecheck.go]: Why export this function? does it even need to be a function?

File: scopecheck/scopecheck.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+	myAuth := Auth(tcclient.ConnectionData{
+		Credentials:  credentials,
+		BaseURL:      "https://auth.taskcluster.net/v1",
```
Comment (jonasfj) [inline scopecheck/scopecheck.go]: this has sane defaults...

File: scopecheck/scopecheck.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+		Authenticate: true,
+	})
+	return &myAuth
+}
+
+//may have duplicate of this method because this method will be written at some point in issue#31
+func (myAuth *Auth) expandScope(payload *SetOfScopes) (*SetOfScopes, error) {
```
Comment (jonasfj) [inline scopecheck/scopecheck.go]: Why not use:
https://godoc.org/github.com/taskcluster/taskcluster-client-go/auth#Auth.ExpandScopes

File: scopecheck/scopecheck.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+	connectionDetails := tcclient.ConnectionData(*myAuth)
+	responseObject, _, err := (&connectionDetails).APICall(payload, "GET", "/scopes/expand", new(SetOfScopes), nil)
+	return responseObject.(*SetOfScopes), err
+}
+
+func (scopecheck) Execute(context extpoints.Context) bool {
+	argv := context.Arguments
+
+	command := argv["scope-check"].(string)
+	provider := extpoints.CommandProviders()[command]
```
Comment (jonasfj) [inline scopecheck/scopecheck.go]: I'm confused?

File: scopecheck/scopecheck.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+	if provider == nil {
+		panic(fmt.Sprintf("Unknown command: %s", command))
+		return false
+	}
+
+	satisfies := argv("<satisfies>")
+	rscope := argv("<rscope>")
+	lscope := argv("<lscope>")
```
Comment (jonasfj) [inline scopecheck/scopecheck.go]: scope1 and scope2

File: scopecheck/scopecheck.go [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+	_, expandedScopes := response.Scopes
+	return expandedScopes.(string), nil
+}
+
+func (scopecheck) Execute(context extpoints.Context) bool {
+	argv := context.Arguments
+
+	command := argv["SCOPECHECK"].(string)
+	provider := extpoints.CommandProviders()[command]
```
Comment (jonasfj) [inline scopecheck/scopecheck.go]: Still no clue what this is good for... Why are you touching CommandProviders()?

What is `argv["SCOPECHECK"].(string)`...

Please take some time and make sure you understand the code and why you have it. Because this doesn't seem to make sense.

File: scopecheck/scopecheck.go [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+	if provider == nil {
+		panic(fmt.Sprintf("Unknown command: %s", command))
+		return false
+	}
+
+	satisfies := argv("<satisfies>")
```
Comment (jonasfj) [inline scopecheck/scopecheck.go]: `argv isn't a function... this won't build, and again, please make sure you understand what the code is supposed to do.

File: scopecheck/scopecheck.go [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+	return resp, err
+}
+
+func (scopecheck) Execute(context extpoints.Context) bool {
+	argv := context.Arguments
+	scope1 := argv["<scope1>"].(string)
+	scope2 := argv["<scope2>"].(string)
+
+	if argv["scope-check"].(bool) {
```
Comment (jonasfj) [inline scopecheck/scopecheck.go]: What is this check good for?

Comment (nanjekyejoannah) [inline scopecheck/scopecheck.go]: After thinking about it..this looks irrelevant...

**Comment by nanjekyejoannah** [comment]
@jonasfj  can you give me a hint on where I may be going wrong and what I should do?

File: scopecheck/scopecheck.go (hunk 11 above)
Comment (jonasfj) [inline scopecheck/scopecheck.go]: Yes

File: subtree_imports.go [hunk 12]
```diff
@@ -3,3 +3,5 @@ package main
 
 import _ "github.com/taskcluster/taskcluster-cli/apis"
 import _ "github.com/taskcluster/taskcluster-cli/extpoints"
+import _ "github.com/taskcluster/taskcluster-cli/scopecheck"
+import _ "github.com/taskcluster/taskcluster-client-go/auth"
```
Comment (jonasfj) [inline subtree_imports.go]: Any idea how that line ended up there?

File: scopecheck/scopecheck.go [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+		return nil, err
+	}
+
+	return resp, err
+}
+
+func (scopecheck) Execute(context extpoints.Context) bool {
+	argv := context.Arguments
+	scope1 := argv["<scope1>"].(string)
+	if argv["satisfies"] == true {
```
Comment (jonasfj) [inline scopecheck/scopecheck.go]: What is the point of this check?

Comment (nanjekyejoannah) [inline scopecheck/scopecheck.go]: For the satisfies keyword.

Comment (nanjekyejoannah) [inline scopecheck/scopecheck.go]: If I just create a variable for this keyword .. I will end up not using it anywhere and I will get a compile error for variable not being used...

Comment (jonasfj) [inline scopecheck/scopecheck.go]: Well, do you need to create a variable?
It's just a syntax separating scope1 and scope2 right... Why would you need to check that it's there...

Comment (nanjekyejoannah) [inline scopecheck/scopecheck.go]: @jonasfj Well to enable use of syntax of a form like `taskcluster scope-check scope1 satisfies scope2` I will introduce logic of this form `satisfies =: argv["<scope1>"].(string)` but I will not use the satisfies variable anywhere in my remaining logic it is to just enable that syntax. so the compiler complains(variable that is not used) but if I just performed this `argv["satisfies"] == true` I avail the syntax and make the compiler happy.....could there be a cleaner way of me achieving this..than this way?..by the way thanks for the review and sorry for the back and forth...

Comment (jonasfj) [inline scopecheck/scopecheck.go]: The arguments have already been parsed when you do: `satisfies =: argv["satisfies"]`, so how could it possibly affect the argument-parsing...

Arguments are parsed based on the `Usage()` string and passed into `Execute(context extpoints.Context)` as a property on context... Consider:
https://github.com/taskcluster/taskcluster-cli/blob/master/main.go#L88-L114

Comment (nanjekyejoannah) [inline scopecheck/scopecheck.go]: when I do `satisfies =: argv["satisfies"]` I am introducing a variable satisfies which the compiler will complain about because I wont use it anywhere later...

File: scopecheck/scopecheck.go [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+	}
+
+	return resp, err
+}
+
+func (scopecheck) Execute(context extpoints.Context) bool {
+	argv := context.Arguments
+	scope1 := argv["<scope1>"].(string)
+	satisfies := argv["satisfies"]
+	_ = satisfies
```
Comment (walac) [inline scopecheck/scopecheck.go]: ?

File: scopecheck/scopecheck.go [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+func checkscopes(scope1 string, scope2 string) string {
+	var secondScope interface{} = scope2
+	if strings.HasPrefix("scope2", "assume:") {
+		//cast scope2 to string array before calling api method expandingScope
+		expandedScope, errs := expandScope(secondScope.([]string))
+		if errs != nil {
+			resp := "Error while trying to expand scopes"
+			return resp
```
Comment (walac) [inline scopecheck/scopecheck.go]: why not just
```go
return "Error while trying to expand scopes"
```
?

File: scopecheck/scopecheck.go [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+	}
+
+	if scope1 == scope2 {
+		return "YES"
+	} else {
+		resp := "NO missing -"
+		return resp
+	}
+
+}
```
Comment (walac) [inline scopecheck/scopecheck.go]: I think of all this might be replaced by [this](https://godoc.org/github.com/taskcluster/taskcluster-base-go/scopes)

**Comment by djmitche** [comment]
Closing due to the age of the PR.  The code is still here if someone wants to pick this up and run with it!
